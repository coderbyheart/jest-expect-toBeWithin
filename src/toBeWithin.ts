import {
    ensureNumbers,
    matcherHint,
    printReceived,
    printExpected
} from 'jest-matcher-utils';

export const toBeWithin = (actual: number, lowerBound: number, upperBound: number) => {
    ensureNumbers(actual, lowerBound, '.toBeWithin');
    ensureNumbers(actual, upperBound, '.toBeWithin');
    const pass = actual >= lowerBound && actual <= upperBound;
    const message = pass
        ? () =>
            matcherHint('.not.toBeWithin') +
            '\n\n' +
            `Expected value not within:\n` +
            `  ${printExpected(lowerBound)} ` +
            `>= n >= ${printExpected(upperBound)}\n` +
            `Received:\n` +
            `  ${printReceived(actual)}`
        : () =>
            matcherHint('.toBeWithin') +
            '\n\n' +
            `Expected value to be within:\n` +
            `  ${printExpected(lowerBound)} ` +
            `>= n >= ${printExpected(upperBound)}\n` +
            `Received:\n` +
            `  ${printReceived(actual)}`;
    return {message, pass};
};
