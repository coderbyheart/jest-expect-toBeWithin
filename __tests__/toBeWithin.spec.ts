import {toBeWithin} from '../src/toBeWithin';

expect.extend({toBeWithin});

describe('.toBeWithin()', () => {
    [
        [1, 2, 3],
        [-Infinity, 1, Infinity],
        [Number.MIN_VALUE, 1, Number.MAX_VALUE],
        [0x11, 0x21, 0x22],
        [0b11, 0b111, 0b1111],
        [0o11, 0o21, 0o22],
        [0.1, 0.19, 0.2],
    ].forEach(([small, within, big]) => {
        it(`{pass: true} expect(${within}).toBeWithin(${small}, ${big})`, () => {
            expect(small).toBeWithin(small, big);
        });

        it(`{pass: false} expect(${big}).toBeWithin(${small}, ${within})`, () => {
            expect(big).not.toBeWithin(small, within);
        });

        it(`throws: [${small}, ${within}, ${big}]`, () => {
            expect(() =>
                expect(within).not.toBeWithin(small, big),
            ).toThrowErrorMatchingSnapshot();

            expect(() =>
                expect(big).toBeWithin(small, within),
            ).toThrowErrorMatchingSnapshot();
        });
    });

    [
        [1, 1, 1],
        [Number.MIN_VALUE, Number.MIN_VALUE, Number.MIN_VALUE],
        [Number.MAX_VALUE, Number.MAX_VALUE, Number.MAX_VALUE],
        [Infinity, Infinity, Infinity],
        [-Infinity, -Infinity, -Infinity]
    ].forEach(([n1, n2, n3]) => {
        test(`equal numbers: [${n1}, ${n2}, ${n3}]`, () => {
            expect(n2).toBeWithin(n1, n3);

            expect(() =>
                expect(n2).not.toBeWithin(n1, n3),
            ).toThrowErrorMatchingSnapshot();

            expect(() =>
                expect(n2).not.toBeWithin(n1, n3),
            ).toThrowErrorMatchingSnapshot();
        });
    });
});
