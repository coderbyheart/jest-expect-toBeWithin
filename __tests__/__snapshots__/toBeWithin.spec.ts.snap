// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.toBeWithin() equal numbers: [-Infinity, -Infinity, -Infinity] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m-Infinity[39m >= n >= [32m-Infinity[39m
Received:
  [31m-Infinity[39m"
`;

exports[`.toBeWithin() equal numbers: [-Infinity, -Infinity, -Infinity] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m-Infinity[39m >= n >= [32m-Infinity[39m
Received:
  [31m-Infinity[39m"
`;

exports[`.toBeWithin() equal numbers: [1, 1, 1] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m1[39m >= n >= [32m1[39m
Received:
  [31m1[39m"
`;

exports[`.toBeWithin() equal numbers: [1, 1, 1] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m1[39m >= n >= [32m1[39m
Received:
  [31m1[39m"
`;

exports[`.toBeWithin() equal numbers: [1.7976931348623157e+308, 1.7976931348623157e+308, 1.7976931348623157e+308] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m1.7976931348623157e+308[39m >= n >= [32m1.7976931348623157e+308[39m
Received:
  [31m1.7976931348623157e+308[39m"
`;

exports[`.toBeWithin() equal numbers: [1.7976931348623157e+308, 1.7976931348623157e+308, 1.7976931348623157e+308] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m1.7976931348623157e+308[39m >= n >= [32m1.7976931348623157e+308[39m
Received:
  [31m1.7976931348623157e+308[39m"
`;

exports[`.toBeWithin() equal numbers: [5e-324, 5e-324, 5e-324] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m5e-324[39m >= n >= [32m5e-324[39m
Received:
  [31m5e-324[39m"
`;

exports[`.toBeWithin() equal numbers: [5e-324, 5e-324, 5e-324] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m5e-324[39m >= n >= [32m5e-324[39m
Received:
  [31m5e-324[39m"
`;

exports[`.toBeWithin() equal numbers: [Infinity, Infinity, Infinity] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32mInfinity[39m >= n >= [32mInfinity[39m
Received:
  [31mInfinity[39m"
`;

exports[`.toBeWithin() equal numbers: [Infinity, Infinity, Infinity] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32mInfinity[39m >= n >= [32mInfinity[39m
Received:
  [31mInfinity[39m"
`;

exports[`.toBeWithin() throws: [-Infinity, 1, Infinity] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m-Infinity[39m >= n >= [32mInfinity[39m
Received:
  [31m1[39m"
`;

exports[`.toBeWithin() throws: [-Infinity, 1, Infinity] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value to be within:
  [32m-Infinity[39m >= n >= [32m1[39m
Received:
  [31mInfinity[39m"
`;

exports[`.toBeWithin() throws: [0.1, 0.19, 0.2] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m0.1[39m >= n >= [32m0.2[39m
Received:
  [31m0.19[39m"
`;

exports[`.toBeWithin() throws: [0.1, 0.19, 0.2] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value to be within:
  [32m0.1[39m >= n >= [32m0.19[39m
Received:
  [31m0.2[39m"
`;

exports[`.toBeWithin() throws: [1, 2, 3] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m1[39m >= n >= [32m3[39m
Received:
  [31m2[39m"
`;

exports[`.toBeWithin() throws: [1, 2, 3] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value to be within:
  [32m1[39m >= n >= [32m2[39m
Received:
  [31m3[39m"
`;

exports[`.toBeWithin() throws: [3, 7, 15] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m3[39m >= n >= [32m15[39m
Received:
  [31m7[39m"
`;

exports[`.toBeWithin() throws: [3, 7, 15] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value to be within:
  [32m3[39m >= n >= [32m7[39m
Received:
  [31m15[39m"
`;

exports[`.toBeWithin() throws: [5e-324, 1, 1.7976931348623157e+308] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m5e-324[39m >= n >= [32m1.7976931348623157e+308[39m
Received:
  [31m1[39m"
`;

exports[`.toBeWithin() throws: [5e-324, 1, 1.7976931348623157e+308] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value to be within:
  [32m5e-324[39m >= n >= [32m1[39m
Received:
  [31m1.7976931348623157e+308[39m"
`;

exports[`.toBeWithin() throws: [9, 17, 18] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m9[39m >= n >= [32m18[39m
Received:
  [31m17[39m"
`;

exports[`.toBeWithin() throws: [9, 17, 18] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value to be within:
  [32m9[39m >= n >= [32m17[39m
Received:
  [31m18[39m"
`;

exports[`.toBeWithin() throws: [17, 33, 34] 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value not within:
  [32m17[39m >= n >= [32m34[39m
Received:
  [31m33[39m"
`;

exports[`.toBeWithin() throws: [17, 33, 34] 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeWithin([22m[32mexpected[39m[2m)[22m

Expected value to be within:
  [32m17[39m >= n >= [32m33[39m
Received:
  [31m34[39m"
`;
